cpp/src/parquet/encoding.cc

```diff
@@ -365,7 +369,7 @@ class PlainEncoder<BooleanType> : public EncoderImpl, virtual public BooleanEnco
       }
       writer.Finish();
     }
-    sink_.UnsafeAdvance(data.length());
+    sink_.UnsafeAdvance(data.length() - data.null_count());
```

Comment:
shouldn't this be n_valid?

Reply:
@emkornfield Parquet encoding stores "valid" value. The invalid value will be marked in rep-levels and def-levels

Reply:
I was referring to the entire value passed into sink_.UnsafeAdvance

This seems like it was incorrect even for all present values because we are advancing the Byte buffer by number of values and not number of bits?

Reply:
(i.e. we seem to be advancing by more bytes then are being reserved in both cases)

Reply:
Yes, the logic here is confusing, but it's right after my change. When input k boolean values.

1. `sink_.Reserve` will only reserve bytes for bits (k).
2. `sink_.UnsafeAdvance` will advance k bytes.

However, when used, `sink_.length()` will only be regarded as bits. So (2) has a bug, but it works here...

I'd like to fix the bug first, and take time to optimize the code later.

Reply:
In that case we need to be reserving k bytes and not k/8.  I think this wasn't caught sooner because columnwriter appears to have a specialization for Boolean values that bypasses this method (e.g. I don't think anything but encoder tests will pass if you put `throw exception(....)` in this method in general.